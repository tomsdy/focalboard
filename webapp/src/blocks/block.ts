// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Utils} from '../utils'

const contentBlockTypes = ['text', 'image', 'divider', 'checkbox'] as const
const blockTypes = [...contentBlockTypes, 'board', 'view', 'card', 'comment'] as const
type ContentBlockTypes = typeof contentBlockTypes[number]
type BlockTypes = typeof blockTypes[number]

interface IBlock {
    readonly id: string
    readonly parentId: string
    readonly rootId: string
    readonly modifiedBy: string

    readonly schema: number
    readonly type: BlockTypes
    readonly title: string
    readonly fields: Readonly<Record<string, any>>

    readonly createAt: number
    readonly updateAt: number
    readonly deleteAt: number
}

interface IMutableBlock extends IBlock {
    id: string
    parentId: string
    rootId: string
    modifiedBy: string

    schema: number
    type: BlockTypes
    title: string
    fields: Record<string, any>

    createAt: number
    updateAt: number
    deleteAt: number
}

class MutableBlock implements IMutableBlock {
    id: string = Utils.createGuid()
    schema: number
    parentId: string
    rootId: string
    modifiedBy: string
    type: BlockTypes
    title: string
    fields: Record<string, any> = {}
    createAt: number = Date.now()
    updateAt = 0
    deleteAt = 0

    constructor(block: any = {}) {
        this.id = block.id || Utils.createGuid()
        this.schema = 1
        this.parentId = block.parentId || ''
        this.rootId = block.rootId || ''
        this.modifiedBy = block.modifiedBy || ''
        this.type = block.type || ''

        // Shallow copy here. Derived classes must make deep copies of their known properties in their constructors.
        this.fields = block.fields ? {...block.fields} : {}

        this.title = block.title || ''

        const now = Date.now()
        this.createAt = block.createAt || now
        this.updateAt = block.updateAt || now
        this.deleteAt = block.deleteAt || 0
    }
}

export type {ContentBlockTypes, BlockTypes}
export {blockTypes, contentBlockTypes, IBlock, IMutableBlock, MutableBlock}
