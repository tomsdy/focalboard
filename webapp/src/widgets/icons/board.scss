@use './standardIcon.scss';

.focalboard-body .BoardIcon {
    @extend .StandardIcon;
    stroke-width: 8px;
}
