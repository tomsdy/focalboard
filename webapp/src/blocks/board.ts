// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Utils} from '../utils'

import {IBlock, MutableBlock} from './block'

type PropertyType = 'text' | 'number' | 'select' | 'multiSelect' | 'date' | 'person' | 'file' | 'checkbox' | 'url' | 'email' | 'phone' | 'createdTime' | 'createdBy' | 'updatedTime' | 'updatedBy'

interface IPropertyOption {
    readonly id: string
    readonly value: string
    readonly color: string
}

interface IMutablePropertyOption {
    id: string
    value: string
    color: string
}

// A template for card properties attached to a board
interface IPropertyTemplate {
    readonly id: string
    readonly name: string
    readonly type: PropertyType
    readonly options: IPropertyOption[]
}

interface IMutablePropertyTemplate extends IPropertyTemplate {
    id: string
    name: string
    type: PropertyType
    options: IMutablePropertyOption[]
}

interface Board extends IBlock {
    readonly icon: string
    readonly description: string
    readonly showDescription: boolean
    readonly isTemplate: boolean
    readonly cardProperties: readonly IPropertyTemplate[]
    duplicate(): MutableBoard
}

class MutableBoard extends MutableBlock implements Board {
    get icon(): string {
        return this.fields.icon as string
    }
    set icon(value: string) {
        this.fields.icon = value
    }

    get description(): string {
        return this.fields.description as string
    }
    set description(value: string) {
        this.fields.description = value
    }

    get showDescription(): boolean {
        return Boolean(this.fields.showDescription)
    }
    set showDescription(value: boolean) {
        this.fields.showDescription = value
    }

    get isTemplate(): boolean {
        return Boolean(this.fields.isTemplate)
    }
    set isTemplate(value: boolean) {
        this.fields.isTemplate = value
    }

    get cardProperties(): IMutablePropertyTemplate[] {
        return this.fields.cardProperties as IPropertyTemplate[]
    }
    set cardProperties(value: IMutablePropertyTemplate[]) {
        this.fields.cardProperties = value
    }

    constructor(block: any = {}) {
        super(block)
        this.type = 'board'

        this.icon = block.fields?.icon || ''
        this.description = block.fields?.description || ''
        if (block.fields?.cardProperties) {
            // Deep clone of card properties and their options
            this.cardProperties = block.fields.cardProperties.map((o: IPropertyTemplate) => {
                return {
                    id: o.id,
                    name: o.name,
                    type: o.type,
                    options: o.options ? o.options.map((option) => ({...option})) : [],
                }
            })
        } else {
            this.cardProperties = []
        }
    }

    duplicate(): MutableBoard {
        const board = new MutableBoard(this)
        board.id = Utils.createGuid()
        return board
    }
}

export {Board, MutableBoard, PropertyType, IPropertyOption, IPropertyTemplate}
