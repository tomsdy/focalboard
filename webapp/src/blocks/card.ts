// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Utils} from '../utils'

import {IBlock, MutableBlock} from './block'

interface Card extends IBlock {
    readonly icon: string
    readonly isTemplate: boolean
    readonly properties: Readonly<Record<string, string>>
    readonly contentOrder: readonly string[]

    duplicate(): MutableCard
}

class MutableCard extends MutableBlock implements Card {
    get icon(): string {
        return this.fields.icon as string
    }
    set icon(value: string) {
        this.fields.icon = value
    }

    get isTemplate(): boolean {
        return Boolean(this.fields.isTemplate)
    }
    set isTemplate(value: boolean) {
        this.fields.isTemplate = value
    }

    get properties(): Record<string, string> {
        return this.fields.properties as Record<string, string>
    }
    set properties(value: Record<string, string>) {
        this.fields.properties = value
    }

    get contentOrder(): string[] {
        return this.fields.contentOrder
    }
    set contentOrder(value: string[]) {
        this.fields.contentOrder = value
    }

    constructor(block: any = {}) {
        super(block)
        this.type = 'card'

        this.icon = block.fields?.icon || ''
        this.properties = {...(block.fields?.properties || {})}
        this.contentOrder = block.fields?.contentOrder?.slice() || []
    }

    duplicate(): MutableCard {
        const card = new MutableCard(this)
        card.id = Utils.createGuid()
        return card
    }
}

export {MutableCard, Card}
