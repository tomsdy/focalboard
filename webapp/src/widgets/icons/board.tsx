// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react'

import './board.scss'

export default function BoardIcon(): JSX.Element {
    return (
        <svg
            xmlns='http://www.w3.org/2000/svg'
            className='BoardIcon Icon'
            viewBox='0 0 100 100'
        >
            <rect
                x='10'
                y='10'
                width='80'
                height='80'
                rx='5'
                ry='5'
            />
            <polyline
                points='28,25 28,55'
                style={{strokeWidth: '15px'}}
            />
            <polyline
                points='50,25 50,70'
                style={{strokeWidth: '15px'}}
            />
            <polyline
                points='72,25 72,45'
                style={{strokeWidth: '15px'}}
            />
        </svg>
    )
}
