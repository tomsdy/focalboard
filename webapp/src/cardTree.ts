// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Block} from './blocks/block';
import {Card} from './blocks/card';
import octoClient from './octoClient';
import {IBlock, IOrderedBlock} from './octoTypes';
import {OctoUtils} from './octoUtils';

class CardTree {
    card: Card
    comments: IBlock[]
    contents: IOrderedBlock[]
    isSynched: boolean

    constructor(private cardId: string) {
    }

    async sync() {
	    const blocks = await octoClient.getSubtree(this.cardId)
        this.rebuild(OctoUtils.hydrateBlocks(blocks))
    }

    private rebuild(blocks: Block[]) {
	    this.card = blocks.find((o) => o.id === this.cardId) as Card

        this.comments = blocks.
            filter((block) => block.type === 'comment').
            sort((a, b) => a.createAt - b.createAt)

	    const contentBlocks = blocks.filter((block) => block.type === 'text' || block.type === 'image') as IOrderedBlock[]
        this.contents = contentBlocks.sort((a, b) => a.order - b.order)

        this.isSynched = true
    }
}

export {CardTree}
