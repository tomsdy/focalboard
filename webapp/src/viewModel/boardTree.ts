// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {IBlock} from '../blocks/block'
import {Board, IPropertyOption, IPropertyTemplate, MutableBoard} from '../blocks/board'
import {BoardView, MutableBoardView} from '../blocks/boardView'
import {Card, MutableCard} from '../blocks/card'
import {CardFilter} from '../cardFilter'
import {Constants} from '../constants'
import octoClient from '../octoClient'
import {OctoUtils} from '../octoUtils'
import {Utils} from '../utils'

type Group = {
    option: IPropertyOption
    cards: Card[]
}

interface BoardTree {
    readonly board: Board
    readonly views: readonly BoardView[]
    readonly cards: readonly Card[]
    readonly cardTemplates: readonly Card[]
    readonly allCards: readonly Card[]
    readonly allBlocks: readonly IBlock[]

    readonly visibleGroups: readonly Group[]
    readonly hiddenGroups: readonly Group[]

    readonly activeView: BoardView
    readonly groupByProperty?: IPropertyTemplate

    getSearchText(): string | undefined
    orderedCards(): Card[]

    copyWithView(viewId: string): BoardTree
    copyWithSearchText(searchText?: string): BoardTree
}

class MutableBoardTree implements BoardTree {
    board: MutableBoard
    views: MutableBoardView[] = []
    cards: MutableCard[] = []
    cardTemplates: MutableCard[] = []

    visibleGroups: Group[] = []
    hiddenGroups: Group[] = []

    activeView!: MutableBoardView
    groupByProperty?: IPropertyTemplate

    private searchText?: string
    allCards: MutableCard[] = []
    get allBlocks(): IBlock[] {
        return [this.board, ...this.views, ...this.allCards, ...this.cardTemplates]
    }

    constructor(board: MutableBoard) {
        this.board = board
    }

    // Factory methods

    static async sync(boardId: string, viewId: string): Promise<BoardTree | undefined> {
        const rawBlocks = await octoClient.getSubtree(boardId)
        const newBoardTree = this.buildTree(boardId, rawBlocks)
        if (newBoardTree) {
            newBoardTree.setActiveView(viewId)
        }
        return newBoardTree
    }

    static incrementalUpdate(boardTree: BoardTree, updatedBlocks: IBlock[]): BoardTree | undefined {
        const relevantBlocks = updatedBlocks.filter((block) => block.deleteAt !== 0 || block.id === boardTree.board.id || block.parentId === boardTree.board.id)
        if (relevantBlocks.length < 1) {
            // No change
            return boardTree
        }
        const rawBlocks = OctoUtils.mergeBlocks(boardTree.allBlocks, relevantBlocks)
        const newBoardTree = this.buildTree(boardTree.board.id, rawBlocks)
        if (newBoardTree && boardTree.activeView) {
            newBoardTree.setActiveView(boardTree.activeView.id)
        }
        return newBoardTree
    }

    private static buildTree(boardId: string, sourceBlocks: readonly IBlock[]): MutableBoardTree | undefined {
        const blocks = OctoUtils.hydrateBlocks(sourceBlocks)
        const board = blocks.find((block) => block.type === 'board' && block.id === boardId) as MutableBoard
        if (!board) {
            return undefined
        }
        const boardTree = new MutableBoardTree(board)
        boardTree.views = blocks.filter((block) => block.type === 'view').
            sort((a, b) => a.title.localeCompare(b.title)) as MutableBoardView[]
        boardTree.allCards = blocks.filter((block) => block.type === 'card' && !(block as Card).isTemplate) as MutableCard[]
        boardTree.cardTemplates = blocks.filter((block) => block.type === 'card' && (block as Card).isTemplate).
            sort((a, b) => a.title.localeCompare(b.title)) as MutableCard[]
        boardTree.cards = []

        boardTree.ensureMinimumSchema()
        return boardTree
    }

    private ensureMinimumSchema(): boolean {
        let didChange = false

        // At least one select property
        const selectProperties = this.board.cardProperties.find((o) => o.type === 'select')
        if (!selectProperties) {
            const newBoard = new MutableBoard(this.board)
            newBoard.rootId = newBoard.id
            const property: IPropertyTemplate = {
                id: Utils.createGuid(),
                name: 'Status',
                type: 'select',
                options: [],
            }
            newBoard.cardProperties.push(property)
            this.board = newBoard
            didChange = true
        }

        // At least one view
        if (this.views.length < 1) {
            const view = new MutableBoardView()
            view.parentId = this.board.id
            view.rootId = this.board.rootId
            view.groupById = this.board.cardProperties.find((o) => o.type === 'select')?.id
            this.views.push(view)
            didChange = true
        }

        if (!this.activeView) {
            this.activeView = this.views[0]
        }

        return didChange
    }

    private setActiveView(viewId?: string): void {
        let view: MutableBoardView | undefined
        if (viewId) {
            view = this.views.find((o) => o.id === viewId)
            if (!view) {
                Utils.logError(`Cannot find BoardView: ${viewId}`)
                view = this.views[0]
            }
        } else {
            // Default to first view
            view = this.views[0]
        }

        this.activeView = view!

        // Fix missing group by (e.g. for new views)
        if (this.activeView.viewType === 'board' && !this.activeView.groupById) {
            this.activeView.groupById = this.board.cardProperties.find((o) => o.type === 'select')?.id
        }

        this.applyFilterSortAndGroup()
    }

    getSearchText(): string | undefined {
        return this.searchText
    }

    private setSearchText(text?: string): void {
        this.searchText = text
        this.applyFilterSortAndGroup()
    }

    private applyFilterSortAndGroup(): void {
        if (!this.activeView) {
            Utils.assertFailure('activeView')
            return
        }
        Utils.assert(this.allCards !== undefined)

        this.cards = this.filterCards(this.allCards) as MutableCard[]
        Utils.assert(this.cards !== undefined)
        this.cards = this.searchFilterCards(this.cards) as MutableCard[]
        Utils.assert(this.cards !== undefined)
        this.cards = this.sortCards(this.cards) as MutableCard[]
        Utils.assert(this.cards !== undefined)

        if (this.activeView.groupById) {
            this.setGroupByProperty(this.activeView.groupById)
        } else {
            Utils.assert(this.activeView.viewType !== 'board')
        }

        Utils.assert(this.cards !== undefined)
    }

    private searchFilterCards(cards: Card[]): Card[] {
        const searchText = this.searchText?.toLocaleLowerCase()
        if (!searchText) {
            return cards.slice()
        }

        return cards.filter((card: Card) => {
            const searchTextInCardTitle: boolean = card.title?.toLocaleLowerCase().includes(searchText)
            if (searchTextInCardTitle) {
                return true
            }

            // Search for text in properties
            const {board} = this
            for (const [propertyId, propertyValue] of Object.entries(card.properties)) {
                // TODO: Refactor to a shared function that returns the display value of a property
                const propertyTemplate = board.cardProperties.find((o) => o.id === propertyId)
                if (propertyTemplate) {
                    if (propertyTemplate.type === 'select') {
                        // Look up the value of the select option
                        const option = propertyTemplate.options.find((o) => o.id === propertyValue)
                        if (option?.value.toLowerCase().includes(searchText)) {
                            return true
                        }
                    } else if (propertyValue.toLowerCase().includes(searchText)) {
                        return true
                    }
                }
            }

            return false
        })
    }

    private setGroupByProperty(propertyId: string) {
        const {board} = this

        let property = board.cardProperties.find((o) => o.id === propertyId)

        // TODO: Handle multi-select
        if (!property || property.type !== 'select') {
            Utils.logError(`this.view.groupById card property not found: ${propertyId}`)
            property = board.cardProperties.find((o) => o.type === 'select')
            Utils.assertValue(property)
        }
        this.groupByProperty = property

        this.groupCards()
    }

    private groupCards() {
        const {activeView, groupByProperty} = this
        if (!activeView || !groupByProperty) {
            Utils.assertFailure('groupCards')
            return
        }

        const unassignedOptionIds = groupByProperty.options.
            filter((o) => !activeView.visibleOptionIds.includes(o.id) && !activeView.hiddenOptionIds.includes(o.id)).
            map((o) => o.id)
        const visibleOptionIds = [...activeView.visibleOptionIds, ...unassignedOptionIds]
        const {hiddenOptionIds} = activeView

        // If the empty group positon is not explicitly specified, make it the first visible column
        if (!activeView.visibleOptionIds.includes('') && !activeView.hiddenOptionIds.includes('')) {
            visibleOptionIds.unshift('')
        }

        this.visibleGroups = this.groupCardsByOptions(visibleOptionIds, groupByProperty)
        this.hiddenGroups = this.groupCardsByOptions(hiddenOptionIds, groupByProperty)
    }

    private groupCardsByOptions(optionIds: string[], groupByProperty: IPropertyTemplate) {
        const groups = []
        for (const optionId of optionIds) {
            if (optionId) {
                const option = groupByProperty.options.find((o) => o.id === optionId)
                if (option) {
                    const cards = this.cards.filter((o) => optionId === o.properties[groupByProperty.id])
                    const group: Group = {
                        option,
                        cards,
                    }
                    groups.push(group)
                } else {
                    Utils.logError(`groupCardsByOptions: Missing option with id: ${optionId}`)
                }
            } else {
                // Empty group
                const emptyGroupCards = this.cards.filter((card) => {
                    const groupByOptionId = card.properties[groupByProperty.id]
                    return !groupByOptionId || !groupByProperty.options.find((option) => option.id === groupByOptionId)
                })
                const group: Group = {
                    option: {id: '', value: `No ${groupByProperty.name}`, color: ''},
                    cards: emptyGroupCards,
                }
                groups.push(group)
            }
        }

        return groups
    }

    private filterCards(cards: MutableCard[]): Card[] {
        const {board} = this
        const filterGroup = this.activeView.filter
        if (!filterGroup) {
            return cards.slice()
        }

        return CardFilter.applyFilterGroup(filterGroup, board.cardProperties, cards)
    }

    private titleOrCreatedOrder(cardA: Card, cardB: Card) {
        const aValue = cardA.title
        const bValue = cardB.title

        if (aValue && bValue) {
            return aValue.localeCompare(bValue)
        }

        // Always put untitled cards at the bottom
        if (aValue && !bValue) {
            return -1
        }
        if (bValue && !aValue) {
            return 1
        }

        // If both cards are untitled, use the create date
        return cardA.createAt - cardB.createAt
    }

    private manualOrder(activeView: BoardView, cardA: Card, cardB: Card) {
        const indexA = activeView.cardOrder.indexOf(cardA.id)
        const indexB = activeView.cardOrder.indexOf(cardB.id)

        if (indexA < 0 && indexB < 0) {
            return this.titleOrCreatedOrder(cardA, cardB)
        } else if (indexA < 0 && indexB >= 0) {
            // If cardA's order is not defined, put it at the end
            return 1
        }
        return indexA - indexB
    }

    private sortCards(cards: Card[]): Card[] {
        const {board, activeView} = this
        if (!activeView) {
            Utils.assertFailure()
            return cards
        }
        const {sortOptions} = activeView

        if (sortOptions.length < 1) {
            Utils.log('Manual sort')
            return cards.sort((a, b) => this.manualOrder(activeView, a, b))
        }

        let sortedCards = cards
        for (const sortOption of sortOptions) {
            if (sortOption.propertyId === Constants.titleColumnId) {
                Utils.log('Sort by title')
                sortedCards = sortedCards.sort((a, b) => {
                    const result = this.titleOrCreatedOrder(a, b)
                    return sortOption.reversed ? -result : result
                })
            } else {
                const sortPropertyId = sortOption.propertyId
                const template = board.cardProperties.find((o) => o.id === sortPropertyId)
                if (!template) {
                    Utils.logError(`Missing template for property id: ${sortPropertyId}`)
                    return sortedCards
                }
                Utils.log(`Sort by property: ${template?.name}`)
                sortedCards = sortedCards.sort((a, b) => {
                    // Always put cards with no titles at the bottom, regardless of sort
                    if (!a.title || !b.title) {
                        return this.titleOrCreatedOrder(a, b)
                    }

                    const aValue = a.properties[sortPropertyId] || ''
                    const bValue = b.properties[sortPropertyId] || ''
                    let result = 0
                    if (template.type === 'select') {
                        // Always put empty values at the bottom
                        if (aValue && !bValue) {
                            return -1
                        }
                        if (bValue && !aValue) {
                            return 1
                        }
                        if (!aValue && !bValue) {
                            return this.titleOrCreatedOrder(a, b)
                        }

                        // Sort by the option order (not alphabetically by value)
                        const aOrder = template.options.findIndex((o) => o.id === aValue)
                        const bOrder = template.options.findIndex((o) => o.id === bValue)

                        result = aOrder - bOrder
                    } else if (template.type === 'number' || template.type === 'date') {
                        // Always put empty values at the bottom
                        if (aValue && !bValue) {
                            return -1
                        }
                        if (bValue && !aValue) {
                            return 1
                        }
                        if (!aValue && !bValue) {
                            return this.titleOrCreatedOrder(a, b)
                        }

                        result = Number(aValue) - Number(bValue)
                    } else if (template.type === 'createdTime') {
                        result = a.createAt - b.createAt
                    } else if (template.type === 'updatedTime') {
                        result = a.updateAt - b.updateAt
                    } else {
                        // Text-based sort

                        // Always put empty values at the bottom
                        if (aValue && !bValue) {
                            return -1
                        }
                        if (bValue && !aValue) {
                            return 1
                        }
                        if (!aValue && !bValue) {
                            return this.titleOrCreatedOrder(a, b)
                        }

                        result = aValue.localeCompare(bValue)
                    }

                    if (result === 0) {
                        // In case of "ties", use the title order
                        result = this.titleOrCreatedOrder(a, b)
                    }

                    return sortOption.reversed ? -result : result
                })
            }
        }

        return sortedCards
    }

    orderedCards(): Card[] {
        const cards: Card[] = []
        for (const group of this.visibleGroups) {
            cards.push(...group.cards)
        }
        for (const group of this.hiddenGroups) {
            cards.push(...group.cards)
        }

        return cards
    }

    private mutableCopy(): MutableBoardTree {
        return MutableBoardTree.buildTree(this.board.id, this.allBlocks)!
    }

    copyWithView(viewId: string): BoardTree {
        const boardTree = this.mutableCopy()
        boardTree.setActiveView(viewId)
        return boardTree
    }

    copyWithSearchText(searchText?: string): BoardTree {
        const boardTree = this.mutableCopy()
        if (this.activeView) {
            boardTree.setActiveView(this.activeView.id)
        }
        boardTree.setSearchText(searchText)
        return boardTree
    }
}

export {MutableBoardTree, BoardTree, Group as BoardTreeGroup}
