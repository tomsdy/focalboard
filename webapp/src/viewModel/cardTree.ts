// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {ContentBlockTypes, contentBlockTypes, IBlock} from '../blocks/block'
import {Card, MutableCard} from '../blocks/card'
import {CommentBlock} from '../blocks/commentBlock'
import {IContentBlock} from '../blocks/contentBlock'
import octoClient from '../octoClient'
import {OctoUtils} from '../octoUtils'

interface CardTree {
    readonly card: Card
    readonly comments: readonly CommentBlock[]
    readonly contents: readonly IContentBlock[]
    readonly allBlocks: readonly IBlock[]
}

class MutableCardTree implements CardTree {
    card: MutableCard
    comments: CommentBlock[] = []
    contents: IContentBlock[] = []

    get allBlocks(): IBlock[] {
        return [this.card, ...this.comments, ...this.contents]
    }

    constructor(card: MutableCard) {
        this.card = card
    }

    // Factory methods

    static async sync(boardId: string): Promise<CardTree | undefined> {
        const rawBlocks = await octoClient.getSubtree(boardId)
        return this.buildTree(boardId, rawBlocks)
    }

    static incrementalUpdate(cardTree: CardTree, updatedBlocks: IBlock[]): CardTree | undefined {
        const relevantBlocks = updatedBlocks.filter((block) => block.deleteAt !== 0 || block.id === cardTree.card.id || block.parentId === cardTree.card.id)
        if (relevantBlocks.length < 1) {
            // No change
            return cardTree
        }
        const rawBlocks = OctoUtils.mergeBlocks(cardTree.allBlocks, relevantBlocks)
        return this.buildTree(cardTree.card.id, rawBlocks)
    }

    private static buildTree(cardId: string, sourceBlocks: readonly IBlock[]): MutableCardTree | undefined {
        const blocks = OctoUtils.hydrateBlocks(sourceBlocks)

        const card = blocks.find((o) => o.type === 'card' && o.id === cardId) as MutableCard
        if (!card) {
            return undefined
        }
        const cardTree = new MutableCardTree(card)
        cardTree.comments = blocks.
            filter((block) => block.type === 'comment').
            sort((a, b) => a.createAt - b.createAt) as CommentBlock[]

        const contentBlocks = blocks.filter((block) => contentBlockTypes.includes(block.type as ContentBlockTypes)) as IContentBlock[]
        cardTree.contents = OctoUtils.getBlockOrder(card.contentOrder, contentBlocks)

        return cardTree
    }

    // private mutableCopy(): MutableCardTree {
    //     return MutableCardTree.buildTree(this.card.id, this.allBlocks)!
    // }
}

export {MutableCardTree, CardTree}
