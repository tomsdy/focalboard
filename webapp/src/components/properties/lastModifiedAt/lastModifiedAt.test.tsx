// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react'
import {render} from '@testing-library/react'

import {CardTree, CardTreeContext, MutableCardTree} from '../../../viewModel/cardTree'

import {MutableCard} from '../../../blocks/card'

import {MutableBoardTree} from '../../../viewModel/boardTree'
import {MutableBoard} from '../../../blocks/board'
import {MutableBlock} from '../../../blocks/block'

import LastModifiedAt from './lastModifiedAt'

describe('componnets/properties/lastModifiedAt', () => {
    test('should match snapshot', () => {
        const cardTree = new MutableCardTree(
            new MutableCard({
                updateAt: Date.parse('15 Jun 2021 16:22:00'),
            }),
        )

        const card = new MutableCard()
        card.id = 'card-id-1'
        card.modifiedBy = 'user-id-1'
        card.updateAt = Date.parse('10 Jun 2021 16:22:00')

        const boardTree = new MutableBoardTree(new MutableBoard([]))
        const block = new MutableBlock()
        block.modifiedBy = 'user-id-1'
        block.parentId = 'card-id-1'
        block.type = 'comment'
        block.updateAt = Date.parse('15 Jun 2021 16:22:00')
        boardTree.rawBlocks.push(block)

        const cardTrees:{ [key: string]: CardTree | undefined } = {}
        cardTrees[card.id] = new MutableCardTree(card)

        const component = (
            <CardTreeContext.Provider value={cardTree}>
                <LastModifiedAt
                    card={card}
                    cardTree={cardTree}
                />
            </CardTreeContext.Provider>
        )

        const {container} = render(component)
        expect(container).toMatchSnapshot()
    })
})
