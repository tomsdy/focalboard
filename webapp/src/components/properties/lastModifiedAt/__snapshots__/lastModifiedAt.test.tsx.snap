// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`componnets/properties/lastModifiedAt should match snapshot 1`] = `
<div>
  <div
    class="LastModifiedAt octo-propertyvalue"
  >
    Thu, Jun 10, 2021 4:22 PM
  </div>
</div>
`;
